flink/src/main/java/org/apache/iceberg/flink/FlinkTypeToType.java

```diff
@@ -64,86 +63,136 @@ private int getNextId() {
+    if (arrayType.getElementType().isNullable()) {
       return Types.ListType.ofOptional(getNextId(), elementType);
     } else {
       return Types.ListType.ofRequired(getNextId(), elementType);
     }
   }
 
   @Override
-  public Type map(KeyValueDataType map, Type keyType, Type valueType) {
+  public Type visit(MultisetType multisetType) {
+    Type elementType = multisetType.getElementType().accept(this);
+    return Types.MapType.ofRequired(getNextId(), getNextId(), elementType, Types.IntegerType.get());
+  }
+
+  @Override
+  public Type visit(MapType mapType) {
     // keys in map are not allowed to be null.
-    if (map.getValueDataType().getLogicalType().isNullable()) {
+    Type keyType = mapType.getKeyType().accept(this);
+    Type valueType = mapType.getValueType().accept(this);
+    if (mapType.getValueType().isNullable()) {
       return Types.MapType.ofOptional(getNextId(), getNextId(), keyType, valueType);
     } else {
       return Types.MapType.ofRequired(getNextId(), getNextId(), keyType, valueType);
     }
   }
 
-  @SuppressWarnings("checkstyle:CyclomaticComplexity")
-  @Override
-  public Type atomic(AtomicDataType type) {
-    LogicalType inner = type.getLogicalType();
-    if (inner instanceof VarCharType ||
-        inner instanceof CharType) {
-      return Types.StringType.get();
-    } else if (inner instanceof BooleanType) {
-      return Types.BooleanType.get();
-    } else if (inner instanceof IntType ||
-        inner instanceof SmallIntType ||
-        inner instanceof TinyIntType) {
-      return Types.IntegerType.get();
-    } else if (inner instanceof BigIntType) {
-      return Types.LongType.get();
-    } else if (inner instanceof VarBinaryType) {
-      return Types.BinaryType.get();
-    } else if (inner instanceof BinaryType) {
-      BinaryType binaryType = (BinaryType) inner;
-      return Types.FixedType.ofLength(binaryType.getLength());
-    } else if (inner instanceof FloatType) {
-      return Types.FloatType.get();
-    } else if (inner instanceof DoubleType) {
-      return Types.DoubleType.get();
-    } else if (inner instanceof DateType) {
-      return Types.DateType.get();
-    } else if (inner instanceof TimeType) {
-      return Types.TimeType.get();
-    } else if (inner instanceof TimestampType) {
-      return Types.TimestampType.withoutZone();
-    } else if (inner instanceof LocalZonedTimestampType) {
-      return Types.TimestampType.withZone();
-    } else if (inner instanceof DecimalType) {
-      DecimalType decimalType = (DecimalType) inner;
-      return Types.DecimalType.of(decimalType.getPrecision(), decimalType.getScale());
-    } else {
-      throw new UnsupportedOperationException("Not a supported type: " + type.toString());
+  @Override
+  public Type visit(RowType rowType) {
+    List<Types.NestedField> newFields = Lists.newArrayListWithExpectedSize(rowType.getFieldCount());
+    boolean isRoot = root == rowType;
+
+    List<Type> types = rowType.getFields().stream()
```

Comment:
Seems here we don't need to loop twice ( the first loop to get `List<Type>` and the next loop to get `List<Types.NestedField>` ). Could be simplified like the following: 

```java
  @Override
  public Type visit(RowType rowType) {
    List<Types.NestedField> newFields = Lists.newArrayListWithExpectedSize(rowType.getFieldCount());
    boolean isRoot = root == rowType;

    for (int i = 0; i < rowType.getFieldCount(); i++) {
      int id = isRoot ? i : getNextId();

      RowType.RowField field = rowType.getFields().get(i);
      String name = field.getName();
      String comment = field.getDescription().orElse(null);
      Type type = field.getType().accept(this);

      if (field.getType().isNullable()) {
        newFields.add(Types.NestedField.optional(id, name, type, comment));
      } else {
        newFields.add(Types.NestedField.required(id, name, type, comment));
      }
    }

    return Types.StructType.of(newFields);
  }
```

Reply:
One thing is :  we may adjust the place to generate field Id for nested types, then we may need to adjust the unit test ..

Reply:
I'd prefer to keep the loop twice. If we need change the generation ID for nested types, I think it is better to change Spark too.

Reply:
I'm OK about the current twice loop here now, let's just keep the consistence id generation with spark here.